### PR TITLE
feat: `marimo export html-wasm` with included assets

### LINKED ISSUE #3116: marimo export html-wasm --mode=(read|edit)
### Description

A new export type `marimo export html-wasm` that exports the app as a runnable wasm notebook. This should output slightly different HTML that runs the notebook in wasm instead of a static notebook.

### Suggested solution

Rather than piggy-backing off `marimo export html`, we should create a new subcommand `html-wasm`, because the CLI args will diverge (--mode vs --include-code, already, and more to come).

The exported wasm should have a warning if the current `window` `protocol` is `file://` because wasm/webworkers will not work in this context. So the warning should be clear that is must be hosted.

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mscolnick**
Closes #3116

## Export to WASM-powered HTML

Export your notebook to a self-contained HTML file that runs using WebAssembly:

```bash
marimo export html-wasm notebook.py -o output_dir --mode run
# Or edit mode
marimo export html-wasm notebook.py -o output_dir --mode edit
```

The exported HTML file will run your notebook using WebAssembly, making it completely self-contained and executable in the browser. This means users can interact with your notebook without needing Python or marimo installed.

Options:

- `--mode`: Choose between `run` (read-only) or `edit` (allows editing)
- `--output`: Directory to save the HTML and required assets

```{admonition} Note
:class: note
The exported file must be served over HTTP to function correctly - it cannot be opened directly from the filesystem (file://).
Your server must also serve the assets in the `assets` directory, next to the HTML file. For this reason, we recommend using the online playground if possible: <https://marimo.app>.
```

### Testing the export

You can test the export by running the following command in the directory containing your notebook:

```bash
python -m http.server
```

This spins up a local HTTP server that serves the HTML and assets. We do not recommend using this for production and instead recommend using a proper HTTP server.

### CODE REVIEW

File: marimo/_cli/export/commands.py [hunk 1]
```diff
@@ -335,7 +341,75 @@ def export_callback(file_path: MarimoPath) -> ExportResult:
...
+@click.option(
+    "-o",
+    "--output",
+    type=str,
+    required=True,
+    help="""Output directory to save the HTML to.""",
+)
+@click.option(
+    "--mode",
+    type=click.Choice(["edit", "run"]),
```
Comment (akshayka) [inline marimo/_cli/export/commands.py:368]: In this context `run` isn't self explanatory. How about `readonly`?

File: marimo/_cli/export/commands.py [hunk 2]
```diff
@@ -335,7 +341,75 @@ def export_callback(file_path: MarimoPath) -> ExportResult:
...
+    help="Mode to run the notebook in.",
```
Comment (akshayka) [inline marimo/_cli/export/commands.py:369]:
```suggestion
    help="Whether the notebook code should be editable or readonly",
```

File: marimo/_cli/export/commands.py [hunk 3]
```diff
@@ -335,7 +341,75 @@ def export_callback(file_path: MarimoPath) -> ExportResult:
...
+    required=True,
+)
+@click.option(
+    "--include-assets/--no-include-assets",
+    default=True,
+    show_default=True,
+    type=bool,
+    help="""
+    Include assets in the exported notebook.
+    These assets are required for the notebook to run in the browser.
```
Comment (akshayka) [inline marimo/_cli/export/commands.py:379]: If the assets are required, why would someone choose to not include them? Can we get rid of this option?

File: marimo/_cli/serve/commands.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Example:
+
+  \b
+  * marimo serve notebook.html
+
+"""
+)
+@click.argument("filename", required=True, type=click.Path(exists=True))
+def serve(filename: str) -> None:
```
Comment (akshayka) [inline marimo/_cli/serve/commands.py:27]: While this is neat, I would prefer to not add `serve` to the CLI right now.

Since this code snippet is so small and doesn't depend on any marimo internals (the port can be hardcoded instead), I would instead suggest adding the code snippet to the docs, or even publishing it as a script/gist on GitHub and recommending it be run with `uv run`

Also, `serve` and `run` are close to synonyms, so the terminology is a bit confusing.

File: marimo/_server/utils.py [hunk 5]
```diff
@@ -44,12 +44,10 @@ def find_free_port(port: int, attempts: int = 100) -> int:
 
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
         try:
-            in_use = sock.connect_ex(("localhost", port)) == 0
-            if not in_use:
-                return port
+            sock.bind(("localhost", port))
+            return port
```
Comment (akshayka) [inline marimo/_server/utils.py:48]: Why this change?

File: docs/guides/exporting.md [hunk 6]
```diff
@@ -132,6 +132,40 @@ You can also use other tools that work with Jupyter notebooks:
...
+- `--mode`: Choose between `run` (read-only) or `edit` (allows editing)
+- `--output`: Directory to save the HTML and required assets
+
+```{admonition} Note
+:class: note
+
+The exported file must be served over HTTP to function correctly - it cannot be opened directly from the filesystem (file://).
+Your server must also serve the assets in the `assets` directory, next to the HTML file. For this reason, we recommend using the online playground if possible: <https://marimo.app>.
+```
+
```
Comment (akshayka) [inline docs/guides/exporting.md:158]: Mention that can be hosted on Github pages?

**Review by akshayka — COMMENTED** [review]
Awesome! Some comments on the CLI.

File: marimo/_cli/serve/commands.py (hunk 4 above)
Comment (akshayka) [reply marimo/_cli/serve/commands.py:27]: If you feel strongly though can be convinced to keep it. Just seems simpler to not include since it doesn't use marimo internals, and is readily accomplished in just a few lines of code or a GitHub gist.

Comment (akshayka) [reply marimo/_cli/serve/commands.py:27]: Could a one liner similar to this work?

https://stackoverflow.com/questions/8058793/single-line-python-webserver

Comment (mscolnick) [reply marimo/_cli/serve/commands.py:27]: i started with the one-liner in the docs, but figured it wasn't clear. i will remove this though

File: marimo/_server/utils.py (hunk 5 above)
Comment (mscolnick) [reply marimo/_server/utils.py]: i reverted. it wasn't working as expected before, but this change made it not work even more

File: docs/guides/exporting.md (hunk 6 above)
Comment (mscolnick) [reply docs/guides/exporting.md:158]: yea i can

File: marimo/_cli/export/commands.py (hunk 3 above)
Comment (mscolnick) [reply marimo/_cli/export/commands.py:379]: they may want to run this across a bunch and not keep copying the assets. i can remove this for now

File: marimo/_cli/export/commands.py (hunk 1 above)
Comment (mscolnick) [reply marimo/_cli/export/commands.py:368]: i was thinking it aligns with `marimo edit` and `marimo run`

File: docs/guides/exporting.md [hunk 7]
```diff
@@ -132,6 +132,79 @@ You can also use other tools that work with Jupyter notebooks:
 - [Quarto](https://quarto.org) - Create beautiful documents, websites, presentations
 - [nbgrader](https://nbgrader.readthedocs.io/) - Grade notebook assignments
 
+## Export to WASM-powered HTML
+
+Export your notebook to a self-contained HTML file that runs using WebAssembly:
+
+```bash
+marimo export html-wasm notebook.py -o output_dir --mode run
```
Comment (akshayka) [inline docs/guides/exporting.md:140]:
```suggestion
# export as readonly, with code locked
marimo export html-wasm notebook.py -o output_dir --mode run
```

File: docs/guides/exporting.md [hunk 8]
```diff
@@ -132,6 +132,79 @@ You can also use other tools that work with Jupyter notebooks:
...
+# Or edit mode
```
Comment (akshayka) [inline docs/guides/exporting.md:141]:
```suggestion
# export as an editable notebook
```

File: docs/guides/exporting.md [hunk 9]
```diff
@@ -132,6 +132,79 @@ You can also use other tools that work with Jupyter notebooks:
...
+You can test the export by running the following command in the directory containing your notebook:
+
+```bash
+cd path/to/output_dir
+python -m http.server
+```
+
+## Deploying to GitHub Pages
+
+You can deploy your marimo WebAssembly notebook to GitHub Pages by following these steps:
```
Comment (akshayka) [inline docs/guides/exporting.md:170]:
```suggestion
You can deploy your WebAssembly marimo notebook to GitHub Pages using the following GitHub Actions workflow:
```

File: marimo/_cli/export/commands.py [hunk 10]
```diff
@@ -335,7 +341,63 @@ def export_callback(file_path: MarimoPath) -> ExportResult:
...
+@click.command(
+    help="""Export a notebook as a WASM-powered standalone HTML file.
+
+Example:
+
+  \b
+  * marimo export html-wasm notebook.py -o notebook.wasm.html
+
+The exported HTML file will run the notebook using WebAssembly, making it
+completely self-contained and executable in the browser.
```
Comment (akshayka) [inline marimo/_cli/export/commands.py:353]:
```suggestion
completely self-contained and executable in the browser. This lets you
share interactive notebooks on the web without setting up
infrastructure to run Python code.

The exported notebook runs using Pyodide, which supports most
but not all Python packages. To learn more, see the Pyodide
documentation.
```

**Review by akshayka — APPROVED** [review]
This is huge! One of our most requested features. 🎉
